client/transaction-pool/src/lib.rs

```diff
@@ -563,160 +570,232 @@ async fn prune_known_txs_for_block<Block: BlockT, Api: graph::ChainApi<Block = B
+							tx_hash,
+							hash,
+							);
 						}
-					}
+						!contains
+					}));
 
-					let extra_pool = pool.clone();
-					// After #5200 lands, this arguably might be moved to the
-					// handler of "all blocks notification".
-					ready_poll.lock().trigger(block_number, move || {
-						Box::new(extra_pool.validated_pool().ready())
+					metrics.report(|metrics| {
+						metrics.block_transactions_resubmitted.inc_by(resubmitted_to_report)
 					});
+				}
 
-					if next_action.revalidate {
-						let hashes = pool.validated_pool().ready().map(|tx| tx.hash).collect();
-						revalidation_queue.revalidate_later(block_number, hashes).await;
-
-						revalidation_strategy.lock().clear();
-					}
+				if let Err(e) = pool
+					.resubmit_at(
+						&id,
+						// These transactions are coming from retracted blocks, we should
+						// simply consider them external.
+						TransactionSource::External,
+						resubmit_transactions,
+					)
+					.await
+				{
+					log::debug!(
+					target: "txpool",
+					"[{:?}] Error re-submitting transactions: {}",
+					id,
+					e,
+					)
 				}
-				.boxed()
+			}
+
+			let extra_pool = pool.clone();
+			// After #5200 lands, this arguably might be moved to the
+			// handler of "all blocks notification".
+			ready_poll
+				.lock()
+				.trigger(block_number, move || Box::new(extra_pool.validated_pool().ready()));
+
+			if next_action.revalidate {
+				let hashes = pool.validated_pool().ready().map(|tx| tx.hash).collect();
+				revalidation_queue.revalidate_later(block_number, hashes).await;
+
+				revalidation_strategy.lock().clear();
+			}
+		}
+		.boxed()
+	}
+}
+
+impl<PoolApi, Block> MaintainedTransactionPool for BasicPool<PoolApi, Block>
+where
+	Block: BlockT,
+	PoolApi: 'static + graph::ChainApi<Block = Block>,
+{
+	fn maintain(&self, event: ChainEvent<Self::Block>) -> Pin<Box<dyn Future<Output = ()> + Send>> {
+		let prev_finalized_block = self.enactment_state.lock().recent_finalized_block;
+		let (proceed, tree_route) = match self
+			.enactment_state
+			.lock()
+			.update_and_check_if_new_enactment_is_valid(&*self.api, &event)
+		{
+			Err(msg) => {
+				log::warn!(target:"txpool", "{msg}");
+				return Box::pin(ready(()))
 			},
-			ChainEvent::Finalized { hash, tree_route } => {
+			Ok(r) => r,
+		};
+
+		let hash = match event {
+			ChainEvent::NewBestBlock { hash, .. } | ChainEvent::Finalized { hash, .. } => hash,
+		};
+
+		let handle_enactment = if proceed {
+			self.handle_enactment(hash, tree_route.clone())
+		} else {
+			Box::pin(ready(()))
+		};
+
+		match event {
+			ChainEvent::NewBestBlock { .. } => handle_enactment,
+			ChainEvent::Finalized { hash, .. } => {
 				let pool = self.pool.clone();
+
+				let enacted = if let Some(prev_finalized_block) = prev_finalized_block {
+					let tree_route = match self.api.tree_route(prev_finalized_block, hash) {
+						Ok(tree_route) => tree_route.expect("tree_route exists. qed."),
```

Comment:
The event already has this tree route attached. No need to calculate this again.

Reply:
[1e3af7e](https://github.com/paritytech/substrate/pull/12305/commits/1e3af7ed662e7386a635f4029078370f646a706d)